### PR TITLE
Optim foreach cleanup for RAdam

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* __->__ #70230
* #70229
* #69982
* #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to RAdam optimizer + cleanup

Differential Revision: [D33767874](https://our.internmc.facebook.com/intern/diff/D33767874)

### CODE REVIEW

File: torch/optim/radam.py [hunk 1]
```diff
@@ -115,14 +121,140 @@ def step(self, closure=None):
...
+    torch._foreach_add_(exp_avgs, grads, alpha=1 - beta1)
+
+    torch._foreach_mul_(exp_avg_sqs, beta2)
+    torch._foreach_addcmul_(exp_avg_sqs, grads, grads, 1 - beta2)
+
+    rect = [math.sqrt((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t))
+            if rho_t > 5 else 0 for rho_t in rho_t_list]
+    unrectified = [0 if rect > 0 else 1. for rect in rect]
+
+    exp_avg_sq_sqrt = torch._foreach_add(torch._foreach_sqrt(exp_avg_sqs), eps)
```
Comment (mikaylagawarecki) [inline torch/optim/radam.py:252]: `eps` was unused in the multitensor radam implementation but used in the single tensor implentation, I believe the correct place to add it is here but I am not super sure whether this omission was intentional, should i be adding this here?

Comment (albanD) [reply torch/optim/radam.py:266]: I wouldn't change the foreach implementation at all in this PR and open an issue about this.

File: torch/optim/radam.py [hunk 2]
```diff
@@ -115,14 +125,150 @@ def step(self, closure=None):
...
+
         return loss
+
+
+def radam(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          state_steps: List[int],
+          foreach: bool = None,
```
Comment (albanD) [inline torch/optim/radam.py:148]: nit torchscript comment.

Comment (albanD) [inline torch/optim/radam.py:148]: nit torchscript comment.

**Review by albanD — COMMENTED** [review]
Looks good.
I think we want to avoid any change in behavior in this PR and move that to later.

**Review by albanD — APPROVED** [review]
LGTM

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767874).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767874).

**Pull request closed without merging**